File: packages/data/lib/src/repositories/local_skill_repository.dart
import 'package:drift/drift.dart';
import 'package:fmf_data/src/database/app_database.dart';
import 'package:fmf_domain/domain.dart';

class LocalSkillRepository implements SkillRepository {
  const LocalSkillRepository(this._db);

  final AppDatabase _db;

  // Static skill catalog — in future, load from remote CMS or local JSON bundle
  static const _skills = [
    Skill(
      id: 'handstand',
      name: 'Handstand',
      description: 'Build balance, body tension, and overhead strength.',
      category: SkillCategory.balance,
    ),
    Skill(
      id: 'pullups',
      name: 'Pull-ups',
      description: 'Develop pulling strength from dead hang to weighted.',
      category: SkillCategory.strength,
    ),
    Skill(
      id: 'handstand_pushups',
      name: 'Handstand Push-ups',
      description: 'Progress from pike press to freestanding HSPU.',
      category: SkillCategory.strength,
    ),
  ];

  @override
  Future<List<Skill>> getSkills() async => _skills;

  @override
  Future<Skill?> getSkillById(String id) async {
    return _skills.where((s) => s.id == id).firstOrNull;
  }

  @override
  Stream<ProgressSnapshot?> watchSkillProgress(String skillId) {
    return (_db.select(_db.skillProgressTable)
          ..where((t) => t.skillId.equals(skillId))
          ..orderBy([(t) => OrderingTerm.desc(t.snapshotDate)])
          ..limit(1))
        .watchSingleOrNull()
        .map(
          (row) => row == null
              ? null
              : ProgressSnapshot(
                  id: row.id,
                  skillId: row.skillId,
                  trackId: row.trackId,
                  snapshotDate: row.snapshotDate,
                  practiceCount: row.practiceCount,
                  lastPracticeDate: row.lastPracticeDate,
                ),
        );
  }
}

File: apps/ios_app/Sources/Data/Repositories/LocalSkillRepository.swift
import Foundation
import GRDB

final class LocalSkillRepository: SkillRepository {
    private let db: AppDatabase

    // Static catalog — future: load from remote CMS or local JSON bundle
    private static let catalog: [Skill] = [
        Skill(
            id: "handstand",
            name: "Handstand",
            description: "Build balance, body tension, and overhead strength.",
            category: .balance,
            prescriptionType: .duration
        ),
        Skill(
            id: "pullups",
            name: "Pull-ups",
            description: "Develop pulling strength and scapular control.",
            category: .strength,
            prescriptionType: .reps
        ),
        Skill(
            id: "handstand_pushups",
            name: "Handstand Push-ups",
            description: "Master overhead pressing strength and balance combined.",
            category: .bodyweight,
            prescriptionType: .reps
        ),
    ]

    init(db: AppDatabase) {
        self.db = db
    }

    func getSkills() async throws -> [Skill] {
        Self.catalog
    }

    func getSkillById(_ id: String) async throws -> Skill? {
        Self.catalog.first { $0.id == id }
    }

    func getProgressSnapshot(skillId: String) async throws -> ProgressSnapshot? {
        try await db.dbWriter.read { db in
            try SkillProgressRecord
                .filter(Column("skillId") == skillId)
                .order(Column("snapshotDate").desc)
                .fetchOne(db)?
                .asDomain
        }
    }

    func saveProgressSnapshot(_ snapshot: ProgressSnapshot) async throws {
        let record = SkillProgressRecord(from: snapshot)
        try await db.dbWriter.write { db in
            try record.save(db)
        }
    }

    func skillProgressStream(skillId: String) -> AsyncStream<ProgressSnapshot?> {
        AsyncStream { continuation in
            let observation = ValueObservation.tracking { db -> ProgressSnapshot? in
                let record = try SkillProgressRecord
                    .filter(Column("skillId") == skillId)
                    .order(Column("snapshotDate").desc)
                    .fetchOne(db)
                return record?.asDomain
            }
            let cancellable = observation.start(
                in: self.db.dbWriter,
                onError: { _ in continuation.yield(nil) },
                onChange: { snapshot in continuation.yield(snapshot) }
            )
            // Box non-Sendable cancellable to satisfy Swift 6 concurrency
            final class CancellableBox: @unchecked Sendable {
                let token: AnyDatabaseCancellable
                init(_ token: AnyDatabaseCancellable) { self.token = token }
            }
            let box = CancellableBox(cancellable)
            continuation.onTermination = { _ in box.token.cancel() }
        }
    }
}
